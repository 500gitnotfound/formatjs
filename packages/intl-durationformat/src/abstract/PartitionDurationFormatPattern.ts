import {invariant} from '@formatjs/ecma402-abstract'
import {TABLE_2} from '../constants'
import {DurationFormat} from '../core'
import {getInternalSlots} from '../get_internal_slots'
import {DurationFormatPart, DurationRecord} from '../types'

export function PartitionDurationFormatPattern(
  df: DurationFormat,
  duration: DurationRecord
): DurationFormatPart[] {
  const result: DurationFormatPart[][] = []
  let done = false
  let separated = false
  const internalSlots = getInternalSlots(df)
  let dataLocale = internalSlots.dataLocale
  const dataLocaleData = DurationFormat.localeData[dataLocale]
  if (!dataLocaleData) {
    throw new TypeError('Invalid locale')
  }
  const numberingSystem = internalSlots.numberingSystem
  const separator = dataLocaleData.digitalFormat[numberingSystem]
  while (!done) {
    let nextStyle
    for (const row of TABLE_2) {
      let value = duration[row.valueField]
      const style = internalSlots[row.styleSlot]
      const display = internalSlots[row.displaySlot]
      const {unit, numberFormatUnit} = row
      const nfOpts = Object.create(null)
      if (
        unit === 'seconds' ||
        unit === 'milliseconds' ||
        unit === 'microseconds'
      ) {
        if (unit === 'seconds') {
          nextStyle = internalSlots.milliseconds
        } else if (unit === 'milliseconds') {
          nextStyle = internalSlots.microseconds
        } else {
          nextStyle = internalSlots.nanoseconds
        }
        if (nextStyle === 'numeric') {
          if (unit === 'seconds') {
            value = value +=
              duration.milliseconds / 1e3 +
              duration.microseconds / 1e6 +
              duration.nanoseconds / 1e9
          } else if (unit === 'milliseconds') {
            value = value +=
              duration.microseconds / 1e3 + duration.nanoseconds / 1e6
          } else {
            value = value += duration.nanoseconds / 1e3
          }
          if (internalSlots.fractionalDigits === undefined) {
            nfOpts.minimumFractionDigits = 0
            nfOpts.maximumFractionDigits = 9
          } else {
            nfOpts.minimumFractionDigits = internalSlots.fractionalDigits
            nfOpts.maximumFractionDigits = internalSlots.fractionalDigits
          }
          nfOpts.roundingMode = 'trunc'
          done = true
        }
      }
      if (value !== 0 || display !== 'auto') {
        nfOpts.numberingSystem = internalSlots.numberingSystem
        if (style === '2-digit') {
          nfOpts.minimumIntegerDigits = 2
        }
        if (style !== '2-digit' && style !== 'numeric') {
          nfOpts.style = 'unit'
          nfOpts.unit = numberFormatUnit
          nfOpts.unitDisplay = style
        }
        const nf = new Intl.NumberFormat(internalSlots.locale, nfOpts)
        let list: DurationFormatPart[]
        if (!separated) {
          list = []
        } else {
          list = result[result.length - 1]
          list.push({
            type: 'literal',
            value: separator,
          })
        }
        let parts = nf.formatToParts(value)
        parts.forEach(part => {
          list.push({
            type: part.type,
            value: part.value,
            unit: numberFormatUnit,
          })
        })
        if (!separated) {
          if (style === '2-digit' || style === 'numeric') {
            separated = true
          }
          result.push(list)
        }
      } else {
        separated = false
      }
    }
  }
  const lfOpts = Object.create(null)
  lfOpts.type = 'unit'
  let listStyle = internalSlots.style
  if (listStyle === 'digital') {
    listStyle = 'short'
  }
  lfOpts.style = listStyle
  const lf = new Intl.ListFormat(internalSlots.locale, lfOpts)
  const strings: string[] = []
  for (const parts of result) {
    let string = ''
    for (const {value} of parts) {
      string += value
    }
    strings.push(string)
  }
  let formatted = lf.formatToParts(strings)
  let resultIndex = 0
  let resultLength = result.length
  let flattened = []
  for (const {type, value} of formatted) {
    if (type === 'element') {
      invariant(resultIndex < resultLength, 'Index out of bounds')
      let parts = result[resultIndex]
      for (const part of parts) {
        flattened.push(part)
      }
      resultIndex++
    } else {
      invariant(type === 'literal', 'Type must be literal')
      flattened.push({
        type: 'literal',
        value,
      })
    }
  }
  return flattened
}
