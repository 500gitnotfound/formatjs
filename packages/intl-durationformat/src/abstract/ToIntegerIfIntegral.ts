import {ToNumber} from '@formatjs/ecma402-abstract'

export function ToIntegerIfIntegral(arg: any) {
  const number = ToNumber(arg)
  if (!Number.isInteger(number)) {
    throw new RangeError(`${arg} is not an integer`)
  }
  return number
}
