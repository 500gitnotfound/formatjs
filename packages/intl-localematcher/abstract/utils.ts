import jsonData from './languageMatching.json'
export const UNICODE_EXTENSION_SEQUENCE_REGEX = /-u(?:-[0-9a-z]{2,8})+/gi

export function invariant(
  condition: boolean,
  message: string,
  Err: any = Error
): asserts condition {
  if (!condition) {
    throw new Err(message)
  }
}

interface LSR {
  language: string
  script: string
  region: string
}

interface LanguageMatchInfo {
  supported: string
  desired: string
  distance: number
  oneway: boolean
}

interface LanguageInfo {
  matches: LanguageMatchInfo[]
  matchVariables: Record<string, string[]>
  paradigmLocales: string[]
}

let PROCESSED_DATA: LanguageInfo | undefined

function processData(): LanguageInfo {
  if (!PROCESSED_DATA) {
    const paradigmLocales = jsonData.supplemental.languageMatching[
      'written-new'
    ][0]?.paradigmLocales?._locales.split(' ') as any
    const matchVariables = jsonData.supplemental.languageMatching[
      'written-new'
    ].slice(1, 5) as any[]
    const data = jsonData.supplemental.languageMatching['written-new'].slice(5)
    const matches = data.map(d => {
      const key = Object.keys(d)[0] as string
      const value = d[key as 'no'] as {
        _desired: string
        _distance: string
        oneway?: string
      }
      return {
        supported: key,
        desired: value._desired,
        distance: +value._distance,
        oneway: value.oneway === 'true' ? true : false,
      }
    }, {})
    PROCESSED_DATA = {
      matches,
      matchVariables: matchVariables.reduce<Record<string, string[]>>(
        (all, d) => {
          const key = Object.keys(d)[0] as string
          const value = d[key]
          all[key.slice(1)] = value._value.split('+')
          return all
        },
        {}
      ),
      paradigmLocales: [
        ...paradigmLocales,
        ...paradigmLocales.map((l: string) =>
          new Intl.Locale(l.replace(/_/g, '-')).maximize().toString()
        ),
      ],
    }
  }

  return PROCESSED_DATA
}

function isMatched(
  locale: LSR,
  languageMatchInfoLocale: string,
  matchVariables: Record<string, string[]>
): boolean {
  const [language, script, region] = languageMatchInfoLocale.split('-')
  let matches = true
  if (region && region[0] === '$') {
    const shouldInclude = region[1] !== '!'
    const matchRegions = shouldInclude
      ? matchVariables[region.slice(1)]
      : matchVariables[region.slice(2)]
    matches ||= matchRegions.includes(locale.region || '') != shouldInclude
  } else if (region === '*' || region === locale.region) {
    matches ||= true
  } else {
    return false
  }

  if (script === '*' || script === locale.script) {
    matches ||= true
  } else {
    return false
  }

  if (language === '*' || language === locale.language) {
    matches ||= true
  } else {
    return false
  }
  return matches
}

function serializeLSR(lsr: LSR): string {
  return [lsr.language, lsr.script, lsr.region].filter(Boolean).join('-')
}

function findMatchingDistanceForLSR(
  desired: LSR,
  supported: LSR,
  data: LanguageInfo
): number {
  for (const d of data.matches) {
    let matches =
      isMatched(desired, d.desired, data.matchVariables) &&
      isMatched(supported, d.supported, data.matchVariables)
    if (!d.oneway && !matches) {
      matches =
        isMatched(desired, d.supported, data.matchVariables) &&
        isMatched(supported, d.desired, data.matchVariables)
    }
    if (matches) {
      console.log('matched', desired, supported, d)
      const distance = d.distance * 10
      if (
        data.paradigmLocales.includes(serializeLSR(desired)) !=
        data.paradigmLocales.includes(serializeLSR(supported))
      ) {
        console.log('matched paradigm locale', desired, supported, d)
        return distance - 1
      }
      return distance
    }
  }
  throw new Error('No matching distance found')
}

export function findMatchingDistance(desired: string, supported: string) {
  const desiredLocale = new Intl.Locale(desired).maximize()
  const supportedLocale = new Intl.Locale(supported).maximize()
  const desiredLSR: LSR = {
    language: desiredLocale.language,
    script: desiredLocale.script || '',
    region: desiredLocale.region || '',
  }
  const supportedLSR: LSR = {
    language: supportedLocale.language,
    script: supportedLocale.script || '',
    region: supportedLocale.region || '',
  }
  let matchingDistance = 0

  const data = processData()
  if (desiredLSR.region !== supportedLSR.region) {
    matchingDistance += findMatchingDistanceForLSR(
      desiredLSR,
      supportedLSR,
      data
    )
  }

  desiredLSR.region = ''
  supportedLSR.region = ''

  if (desiredLSR.script !== supportedLSR.script) {
    matchingDistance += findMatchingDistanceForLSR(
      desiredLSR,
      supportedLSR,
      data
    )
  }

  desiredLSR.script = ''
  supportedLSR.script = ''

  if (desiredLSR.language !== supportedLSR.language) {
    matchingDistance += findMatchingDistanceForLSR(
      desiredLSR,
      supportedLSR,
      data
    )
  }

  return matchingDistance
}

export function findBestMatch(
  desired: string,
  supportedLocales: string[]
): string | undefined {
  let bestMatch = undefined
  let lowestDistance = Infinity
  supportedLocales.forEach(supported => {
    const distance = findMatchingDistance(desired, supported)
    if (distance < lowestDistance) {
      lowestDistance = distance
      bestMatch = supported
    }
  })

  return bestMatch
}
