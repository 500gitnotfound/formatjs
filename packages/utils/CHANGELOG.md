# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 1.1.0 (2024-07-05)

### Features

* **@formatjs/utils:** introduce @formatjs/utils that has a collection of useful intl-relative utilities ([9614b18](https://github.com/formatjs/formatjs/commit/9614b1876503251769f1dbf5de508e80639147e0)) - by @longlho
