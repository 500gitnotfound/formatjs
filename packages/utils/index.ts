export {canonicalizeCountryCode} from './src/countryCodes'
export {defaultCurrency} from './src/defaultCurrency'
