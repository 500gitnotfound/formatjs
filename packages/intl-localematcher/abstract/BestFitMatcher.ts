import {LookupMatcherResult} from './types'
import {UNICODE_EXTENSION_SEQUENCE_REGEX, findBestMatch} from './utils'

/**
 * https://tc39.es/ecma402/#sec-bestfitmatcher
 * @param availableLocales
 * @param requestedLocales
 * @param getDefaultLocale
 */
export function BestFitMatcher(
  availableLocales: readonly string[],
  requestedLocales: readonly string[],
  getDefaultLocale: () => string
): LookupMatcherResult {
  let foundLocale: string | undefined
  let extension: string | undefined
  for (const l of requestedLocales) {
    const matchedLocale = findBestMatch(
      l.replace(UNICODE_EXTENSION_SEQUENCE_REGEX, ''),
      availableLocales
    )
    if (matchedLocale) {
      foundLocale = matchedLocale
      const noExtensionLocale = l.replace(UNICODE_EXTENSION_SEQUENCE_REGEX, '')
      extension = l.slice(noExtensionLocale.length, l.length) || undefined
      break
    }
  }
  if (!foundLocale) {
    return {locale: getDefaultLocale()}
  }

  return {
    locale: foundLocale,
    extension,
  }
}
