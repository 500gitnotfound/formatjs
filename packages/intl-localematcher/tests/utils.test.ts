import {findBestMatch, findMatchingDistance} from '../abstract/utils'
test('findMatchingDistance', () => {
  expect(findMatchingDistance('es-CO', 'es')).toBe(49)
  expect(findMatchingDistance('es-CO', 'es-419')).toBe(39)
  expect(findMatchingDistance('en-US', 'en-US')).toBe(0)
  expect(findMatchingDistance('zh-TW', 'zh-Hant')).toBe(0)
  expect(findMatchingDistance('zh-HK', 'zh-NO')).toBe(540)
  expect(findMatchingDistance('zh-HK', 'zh-Hant')).toBe(40)
  expect(findMatchingDistance('vi', 'fr')).toBe(840)
  expect(findMatchingDistance('sr-Latn-BA', 'sr-Latn-RS')).toBe(40)
  expect(findMatchingDistance('sr-Latn-BA', 'bs-Latn-BA')).toBe(800)
})
test('findBestMatch', () => {
  expect(findBestMatch(['en-US'], ['en-US'])).toMatchSnapshot()
  expect(findBestMatch(['sr-Latn-BA'], ['bs', 'sh'])).toMatchSnapshot()
  expect(findBestMatch(['fr-XX'], ['fr', 'en'])).toMatchSnapshot()
  expect(findBestMatch(['zh-TW'], ['zh', 'zh-Hant'])).toMatchSnapshot()
  expect(findBestMatch(['th-u-ca-gregory'], ['th'])).toMatchSnapshot()

  expect(findBestMatch(['es-co'], ['en', 'es-419'])).toMatchSnapshot()
  expect(findBestMatch(['es-co'], ['en', 'es', 'es-419'])).toMatchSnapshot()
  expect(findBestMatch(['pt-mz'], ['pt-PT', 'pt-BR'])).toMatchSnapshot()
  expect(
    findBestMatch(['de-DE', 'fr-FR'], ['en', 'en-US', 'fr-FR'])
  ).toMatchSnapshot()
})
