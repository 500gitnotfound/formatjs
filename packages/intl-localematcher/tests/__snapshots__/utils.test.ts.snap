// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findBestMatch 1`] = `
{
  "distances": {
    "en-US": {
      "en-US": 0,
    },
  },
  "matchedDesiredLocale": "en-US",
  "matchedSupportedLocale": "en-US",
}
`;

exports[`findBestMatch 2`] = `
{
  "distances": {
    "sr-Latn-BA": {
      "bs": 800,
      "sh": 41,
    },
  },
  "matchedDesiredLocale": "sr-Latn-BA",
  "matchedSupportedLocale": "sh",
}
`;

exports[`findBestMatch 3`] = `
{
  "distances": {
    "fr-XX": {
      "en": 839,
      "fr": 40,
    },
  },
  "matchedDesiredLocale": "fr-XX",
  "matchedSupportedLocale": "fr",
}
`;

exports[`findBestMatch 4`] = `
{
  "distances": {
    "zh-TW": {
      "zh": 540,
      "zh-Hant": 1,
    },
  },
  "matchedDesiredLocale": "zh-TW",
  "matchedSupportedLocale": "zh-Hant",
}
`;

exports[`findBestMatch 5`] = `
{
  "distances": {
    "th-u-ca-gregory": {
      "th": 0,
    },
  },
  "matchedDesiredLocale": "th-u-ca-gregory",
  "matchedSupportedLocale": "th",
}
`;

exports[`findBestMatch 6`] = `
{
  "distances": {
    "es-co": {
      "en": 839,
      "es-419": 40,
    },
  },
  "matchedDesiredLocale": "es-co",
  "matchedSupportedLocale": "es-419",
}
`;

exports[`findBestMatch 7`] = `
{
  "distances": {
    "es-co": {
      "en": 839,
      "es": 50,
      "es-419": 41,
    },
  },
  "matchedDesiredLocale": "es-co",
  "matchedSupportedLocale": "es-419",
}
`;

exports[`findBestMatch 8`] = `
{
  "distances": {
    "pt-mz": {
      "pt-BR": 50,
      "pt-PT": 39,
    },
  },
  "matchedDesiredLocale": "pt-mz",
  "matchedSupportedLocale": "pt-PT",
}
`;

exports[`findBestMatch 9`] = `
{
  "distances": {
    "de-DE": {
      "en": 838,
      "en-US": 839,
      "fr-FR": 842,
    },
    "fr-FR": {
      "en": 878,
      "en-US": 879,
      "fr-FR": 42,
    },
  },
  "matchedDesiredLocale": "fr-FR",
  "matchedSupportedLocale": "fr-FR",
}
`;
